### ISSUE TITLE
Competency reporting to student dashboard

### ISSUE THREAD

**Issue by OECDPILA** [pila-project/pila-platform#141]
(High priority)

- [x] 1. In the science activity, the student dashboard screen needs to be widened (especially given Thai may be longer). I also answered the questions incorrectly so I shouldn't have received so many points on justify and conclude
![image](https://github.com/pila-project/pila-platform/assets/145113767/4d35aaf2-0ae9-4fd0-9b83-62d936005253)

- [x] 2. In paper planes science activity, I do not see the 'justify' scoring (but this may be due to space? or does it not show up at all if it is 0/X? It would be good if it still did show up, even when student got 0): 
![image](https://github.com/pila-project/pila-platform/assets/145113767/10fd4249-656b-46fa-875f-bd8fb8e1ae63)

- [x] 3. In Chirpy's Adventure (level 1 and 2), even when I pick up only one or two leaves and lose three lives right away, I still get full score (to clarify, it should show the competency information from that trial, rather than best score to avoid confusion by students BUT in the teacher dashboard, it will save "best" trial - perhaps this is where the issue lies for certain games?):
![image](https://github.com/pila-project/pila-platform/assets/145113767/635c5857-e3a8-4d97-8e5b-7d33297a754b)
![image](https://github.com/pila-project/pila-platform/assets/145113767/f234cf44-773c-4806-8d9c-68752421fe58)

- [x] 4. For Cleaning up our oceans, I still receive full points for subtraction and comparison even if I just pick up one piece of trash and bring it to the boat each time. I confirmed with Sonia that we indeed implemented a scoring rule for subtraction and comparison that is dependent on at least two pieces of trash being picked up together. It works for Level 2.

**Comment by jasonford** [comment]
Task 1: @OECDPILA we can widen the modal and make sure the text stays on 1 line. (points issue handled in task 3)

Task 2: @stephanemagnenat can you set the [0, N] state for unachieved competencies on your side? The dashboard itself should probably stay pretty naive in rendering whatever it's given.

Task 3:@OECD + @stephanemagnenat, It would be nice to have the "latest run" competency info passed on Agent.close for this case but since it sounds like it might be difficult to do that from our previous discussions, so I'll propose a possible easier fix (and after writing it, I think the easier fix might actually be better than passing on Agent .close...): how about we store latest run info in pila/competencies/gameId/latest and keep the "best run" game info in pila/competencies/gameId as it is? In the future the knowlearning core can do all that "historical max" stuff and let you just track 1 state, but for now I think this could be the easiest way to get what we want.

Task 4: @stephanemagnenat is this something to do on the Candli side? If not, it's probably handled by task 3 as well.

**Comment by stephanemagnenat** [comment]
@OECDPILA when reporting such issue, could you add numbers to item or sub-title? I'm now confused to which "task" Jason refers to, thank you! (jetlag doesn't help to parse issues ;-)).

Task 2+3: the agreed approach in Paris is that we always set the max of the points, and this dialog will show that max as well. So it works as intended. Now if we intend something else, we can change, and I fully agree with your proposed approach @jasonford, but I'm not sure that now is the right time to do some refactoring/feature change there, as there is a risk that there are unexpected negative consequences to such a change. I'll see if I find time to investigate that, but  cannot promise anything, and I will priorities safety over features. @jasonford I am not sure that `pila/competencies/gameId/latest` is a good namespace because it breaks the invariant that all entries of `pila/competencies/gameId/` are item scores, which will complexify the parsing code. What about instead using a sibling namespace, for example `pila/latest_competencies/gameId/`?

Task 4: most likely this is something on our side, I'll see with Sonia.

**Comment by jasonford** [comment]
@stephanemagnenat sibling namespace sounds like a good idea!

@OECDPILA could task 4 have been due to older competency achievements, or was it for sure a fresh take?

For task 2 and 3 I agree prioritizing safety over features is a good policy for right now. @OECDPILA, fine to run without them?

**Comment by OECDPILA** [comment]
Sure @stephanemagnenat  - updated for clarity.

For task 2-3, @jasonford @stephanemagnenat  that is fine to run without new features but I think this will be quite confusing for students if it continues to report their "best" run without adding at the least something in the dashboard screen to clarify this. If I were a student and I know I did worse than I did in the previous run but then I get the same feedback, I would think it was a bug in the system, which introduces distrust. Would it be possible to add "Your personal best" at the bottom of all screens? Happy for other suggestions too.

For task 4 - yes, I did multiple fresh takes in new assignments.

**Comment by stephanemagnenat** [comment]
@jasonford I am now writing in parallel to both states: `pila/competencies/GAME_ID` and `pila/latest_competencies/GAME_ID` (except `general:attempts` as this one only makes sense for the persistent state), can you clear `latest_competencies` before starting the game or tell me how to do this clearing properly using the agent API? (I am worried of breaking the proxy otherwise)

**Comment by stephanemagnenat** [comment]
I changed the icon as well.

**Comment by stephanemagnenat** [comment]
@OECDPILA for point 4 I just checked the source of cleaning the ocean level 1 and the check is done for having at least two pieces of trash, so am I confused. Are you sure that you are not seeing an effect of looking at the best score ever? Could there be a bug where the competencies state is not scoped by the assignment?

**Comment by OECDPILA** [comment]
I just created a new assignment and tried the oceans level 1, picking only one piece of trash up at a time and delivering it back to the boat. This is what I see on the dashboard:

![image](https://github.com/pila-project/pila-platform/assets/145113767/a23c9bbe-7211-477a-9c33-d3f97e0c174d)

**Comment by jasonford** [comment]
@OECDPILA could this be the because of old competency data from past attempts? I still need to update to using latest state reports Stephane implemented

@stephanemagnenat I think I can handle all the open issues on my end now based on your updates. Thank you!

**Comment by jasonford** [comment]
@stephanemagnenat + @OECDPILA, everything should be wired in to use latest_competencies from our side now.

I noticed using our embed tool that the first chirpy bird is requesting the pila/latest_competencies/gameId scope, but is not writing into it: https://embed.knowlearning.systems/7d7c4c3b-e7a1-4588-bb78-a859fffccfac

If the switch to updating that state becomes too burdensome, I can always swap back to drawing from pila/competencies/gameId and add a little "Your best score" notification on the dashboard with no sweat.

**Comment by jasonford** [comment]
Looks to be the same for Oceans as well. If you log in to the embed tool here: https://embed.knowlearning.systems/ you can test any link you like!

Note for @OECDPILA: the embed tool will work for sites like cand.li and betty, since they don't require any back end configuration to work. Sites like x.pilaproject.org and karel and create require back ends to be linked up and so won't work in that tool as expected

**Comment by stephanemagnenat** [comment]
@jasonford thank you for spotting the problem, I made a typo mistake! It should be fixed now, sorry for the noise!
And thanks for the reminder tip about the embed page, it does help for testing!

**Comment by jasonford** [comment]
@stephanemagnenat thanks for adjusting! I was testing using the embed page and it looks like the pila/latest_competencies/GAME_ID scope is differentiated from the pila/competencies/GAME_ID scope, but the pila/latest_competencies/GAME_ID is still remembering the highest score. Looks like we're really close to closing though. Thanks for the last minute adjustments!

**Comment by stephanemagnenat** [comment]
@jasonford, I'm not clearing `pila/latest_competencies/GAME_ID`, that's why I was asking yesterday:

> can you clear latest_competencies before starting the game or tell me how to do this clearing properly using the agent API? (I am worried of breaking the proxy otherwise)

I am not sure how to properly do the clearing myself, and I thought that maybe it is easier to do it on the PILA side, to avoid having each embedded application having to do it. But I'm happy to do it on the Candli side, if you tell me what is the right way to do it without breaking the proxy.

**Comment by jasonford** [comment]
@stephanemagnenat cool, I can do that! Sorry, missed that comment after our last dev cycle.

**Comment by stephanemagnenat** [comment]
No problem, thank you very much for doing it!

**Comment by jasonford** [comment]
@stephanemagnenat I have implemented latest_competency clearing successfully! It looks like there is a bug left though that I need your help with (I'm tracking/viewing states with the embed site link to chirpy bird 1: https://embed.knowlearning.systems/7d7c4c3b-e7a1-4588-bb78-a859fffccfac ):

Working case:
1. Load and start game (latest competency data cleared before game makes request to load that state)
2. Play game and see scores increasing as expected, finish game (lose all lives or complete game)
3. Reload the tab and see current top score and the wiped out latest_competency data as expected
4. Play again, and see latest score increasing as expected

Bug case:
1. Load and start game
2. Play game and see scores increasing as expected
3. Press Retry on Chirpy end screen
4. Play game and see score jumps to max score given on last play

I think if this is worked out, I think it has a good chance at being our last bug! Thanks for your help with all this back and forth!

**Comment by stephanemagnenat** [comment]
Hi @jasonford, thank you very much for clearing the latest competency state!

So normally using `pila-play`, the end game screen should never appear when embedded in PILA because at the same time the screen is shown, `close` is sent to the agent. However I discovered a merge error which lead to `success: true` not being given as parameter of `agent.close()` in that case. I assume that is the source of the problem. I have corrected, so I assume that now it should work. Thank you for all the testing and spotting such mistakes!

**Comment by jasonford** [comment]
@stephanemagnenat glad to help!

We ended up deciding to handle the close even as more of a "game done" event in the sequence wrapper so we could let users retry without having to re-load the game. We're clearing "latest_competency" state on game load as well as the end of a run (on Agent.close from your side) to make sure it is fresh for every retry. It is an option to just move users on to the next sequence item and force them to go back and reload the Cand.li item to retry, but I think it'll be nicer for them to use your "retry" screen and avoid reloading.

Is there an easy way on your end to patch the "latest_competency" writing behavior? The reproduction above has behaved reliable for me, so I hope it might help isolate a clear spot in the code. If its too much, I think we can find an acceptable fallback with messaging on the dashboard though.

Thanks again!

**Comment by stephanemagnenat** [comment]
> Is there an easy way on your end to patch the "latest_competency" writing behavior?

I am not sure to understand the question. I understand that you wish that Candli clears `latest_competency` on game over, but that feels inconsistent with the statement that you already clear it on your side on `Agent.close()` (which is sent on game over). So I am confused, what is not working as intended?

Note that if needed, clearing `latest_competency` on game over on the Candli side is possible but I would still need you to tell me how to do that without destroying the proxy object ;-) I assume that setting the state to `{}` would destroy the proxy, wouldn't it?

Also, a question for clarification: now Candli in PILA does show end screens, right? If so, I need to tell Sonia that they need to be translated and not replaced by simple generic game over. Can you confirm that please?

**Comment by jasonford** [comment]
@stephanemagnenat happy to clarify! We do clear the state on Agent.close(), but I noticed the bug mentioned above and think that perhaps there is some local cache of the latest competency state on the Candli side that needs clearing. Were you able to reproduce that bug? For clarity, I'll add a bit more detail:

1) load the game on https://embed.knowlearning.systems/7d7c4c3b-e7a1-4588-bb78-a859fffccfac and press play. You can exit full screen and see the 2 scopes that Candli is writing to on the right side.
2) play the game and get points a few times, then lose.
3) see that the state for latest_competency has been cleared (we do that on our side)
4) press retry
5) score points 1 time and see that the score immediately jumps back to the score achieved on the last run

Also worth noting is that if you reload and play again, the latest_competency works as expected, but the bug presents itself again on retry.

As you mentioned, we do clear the state on Agent.close so no need to do it on the Cand.li side, unless that perhaps helps fixing the above behavior, then it might be the cleanest solution. You're correct a simple ```Object.keys(state).forEach(k => delete state[k])``` will do as far as persistence on our end is concerned.

It might be worth mentioning that a possible way to have the state bound 2 ways locally is to:
```js
Agent.watch(scope, ({ state }) => myLocalReference.forState = state)
```

There are definitely downsides to doing 2 way bindings for state like that, but thought I would mention in case it helps with a quick fix.

> Also, a question for clarification: now Candli in PILA does show end screens, right? If so, I need to tell Sonia that they need to be translated and not replaced by simple generic game over. Can you confirm that please?

I think a generic candli based game-over screen is good behavior for embedded mode, so long as there is always a "retry" button.

**Comment by stephanemagnenat** [comment]
Hi @jasonford thank you for the clarification, it is now clear. I can indeed reproduce the problem. There should be caching on our side, but it seems that somehow the variables survive. I am investigating.

**Comment by stephanemagnenat** [comment]
I confirm that the state of the Candli side is properly cleared, so somehow there is some caching ongoing somewhere else on the Candli side. I will investigate that later today. Thank you for providing a very clear reproducible test case!

**Comment by stephanemagnenat** [comment]
Hi @jasonford, so, I have added debug info at each report:
- `REPORT: BEFORE: competencies`: the competencies state just before reporting,
- `REPORT: BEFORE: latestCompetencies`: the latestCompetencies state just before reporting,
- `REPORT: PASSED: vars`: the new data passed by Candli, can be used to exclude a caching problem on the Candli side.

Here are my observations:
- A fresh re-fetch of the latest competencies (using `await klBrowserAgent.state('pila/latest_competencies/GAME_ID');`) after a retry correctly shows the empty object.
- After a retry, the (original proxy) state `latest_competencies` come already filled with its values from the previous run.
- Candli always passes the correct updates (as seen by the `vars` array in the debug log), so it seems that it is not a caching issue on the Candli side.

So it looks to me that the clear in the back-end somehow does not propagate to a state proxy already fetched. Could you thus investigate on your side using the nice test-bed you set up and my extra debug info? If it seems too hard to fix, I'm happy to add the manual clearing in addition but I guess it is better to first try to understand the root cause of the problem.

**Comment by jasonford** [comment]
@stephanemagnenat, yeah agreed it's best to find the root! I think it might be a caching issue that understandably requires clarification on api expectations:

The returned proxy for a state already fetched is not automatically updated when updated in other places. For most cases, our system assumes you want to manage a local copy after a request so you won't get off guard by updates outside the scope of your  program. If you do want to keep your local copy in sync with remote updates, you can do something like:

```js
Agent.watch(scope, ({ state }) => myLocalReference.forState = state)
```

Additionally, and maybe easier, if you fetch the state object freshly with a new ```await Agent.state('pila/latest_competencies/...')``` call, It will be the most recent global version.

But in this particular case, it might be easiest to clear the 'latest_competencies' at the beginning of each new game on your side, since that will update your local cached copy too (making the clear on our end unnecessary).

Let me know if this could use any more clarification or if any of these approaches sounds good to you. This is certainly a central area we need to find good ways of clarifying for api consumers, Thanks for going through these dev cycles with me!

**Comment by stephanemagnenat** [comment]
Ok, I'm now clearing the `latest_competencies` state with the method you suggested, just after fetching it and at each retry button click. I tested and it seems to work. I let you confirm and close the issue? Also, you should be able to remove the clearing of `latest_competencies` from your side now.

**Comment by jasonford** [comment]
Thanks @stephanemagnenat. Confirmed that the state is updating as expected now. Thanks again for all the dev cycles!
